src/values.rs

```diff
@@ -234,56 +234,44 @@ impl JitValue {
                         let elem_ty = registry.get_type(&info.ty)?;
                         let elem_layout = elem_ty.layout(registry)?.pad_to_align();
 
-                        let ptr: *mut NonNull<()> =
-                            libc::malloc(elem_layout.size() * data.len()).cast();
+                        let ptr: *mut () = libc::malloc(elem_layout.size() * data.len()).cast();
                         let len: u32 = data.len().try_into().unwrap();
 
                         for elem in data {
                             let elem = elem.to_jit(arena, registry, &info.ty)?;
 
                             std::ptr::copy_nonoverlapping(
                                 elem.cast::<u8>().as_ptr(),
-                                NonNull::new(
-                                    ((NonNull::new_unchecked(ptr).as_ptr() as usize)
-                                        + len as usize * elem_layout.size())
-                                        as *mut u8,
-                                )
-                                .unwrap()
-                                .cast()
-                                .as_ptr(),
+                                ptr.byte_add(len as usize * elem_layout.size()).cast::<u8>(),
                                 elem_layout.size(),
                             );
                         }
 
-                        let target = arena.alloc_layout(
-                            Layout::new::<*mut NonNull<()>>()
-                                .extend(Layout::new::<u32>())?
-                                .0
-                                .extend(Layout::new::<u32>())?
-                                .0,
-                        );
+                        let target = arena
+                            .alloc_layout(
+                                Layout::new::<*mut ()>() // ptr
+                                    .extend(Layout::new::<u32>())? // start
+                                    .0
+                                    .extend(Layout::new::<u32>())? // end
+                                    .0
+                                    .extend(Layout::new::<u32>())? // capacity
+                                    .0
```

Comment:
This changes the layout from the original code (adds one field). Are you sure that both are correct?

Reply:
yes, the previous was wrong as stated in the pr comment

Reply:
I can confirm that it's correct as it is now (the array now has 4 fields instead of 3).